tools/clang/lib/CodeGen/CGExpr.cpp

```diff
@@ -1791,11 +1792,36 @@ void CodeGenFunction::EmitStoreThroughExtVectorComponentLValue(RValue Src,
       }
       Builder.CreateStore(Vec, VecDstPtr);
     } else {
-      for (unsigned i = 0; i < VecTy->getVectorNumElements(); i++) {
-        if (llvm::Constant *Elt = Elts->getAggregateElement(i)) {
-          llvm::Value *EltGEP = Builder.CreateGEP(VecDstPtr, {Zero, Elt});
-          llvm::Value *SrcElt = Builder.CreateExtractElement(SrcVal, i);
-          Builder.CreateStore(SrcElt, EltGEP);
+      // If the vector pointer comes from a call inst, assume it's a subscript.
+      // Do a load and then a store to avoid generating multiple stores.
+      if (llvm::CallInst *CI = llvm::dyn_cast<llvm::CallInst>(VecDstPtr)) {
+        llvm::Value *vec = llvm::UndefValue::get(VecTy);
+        llvm::Value *old_vec = Builder.CreateLoad(VecDstPtr);
+
+        llvm::SmallVector<bool, 4> Stored(VecTy->getVectorNumElements());
+        for (unsigned i = 0; i < VecTy->getVectorNumElements(); i++) {
+          if (llvm::Constant *Elt = Elts->getAggregateElement(i)) {
+            llvm::Value *SrcElt = Builder.CreateExtractElement(SrcVal, i);
+            vec = Builder.CreateInsertElement(vec, SrcElt, i);
```

Comment:
Shouldn't this be inserting into the `Elt` index instead of `i`?

Can we make it simpler by just threding the insert through the old vector?

Like
```
vec = old_vec

vec = InsertElement(vec, )
```

Then any elements that are not modified would already be correct.

Reply:
If the `Elt` variable is not to be used, the conditional should change to just testing the output of `getAggregateElement()`. Clang and MSVC don't care, but GCC does and I occasionally return to that neglected child of mine to fix its many complaints. 😆

Reply:
Yeah that was a mistake. Fixed now.